### PR TITLE
Add support for global sequence processing to the "ordered" extension in Java SDK

### PR DESCRIPTION
**Pull request by slilichenko**
Global sequence processing is used to ensure that events for a given key are only processed when it's guaranteed that they all the elements for this particular have been received. 

Consider a PCollection which contains these event tuples (first element of the tuple is the global sequence number): 
[1, key1, data], [2, key2, data], [3, key1, data], [4, key1, data], [5, key2, data], [7, key2, data]. Elements for key1 must be processed in the following order: 1, 3, 4. Elements for key2 must be processed in the following order: 2, 5. Event with sequence 7 can't be processed because there is a missing sequence 6.

The approach used to implement ordered processing in the presence of global sequencing:

- Generate (periodically for streaming pipelines, once for batch) the side input which contains the maximum contiguous range of sequences across all keys and the maximum timestamp of events in that range.
- Use this side input by the DoFns which process events to a) store in the per-key processing state the latest maximum range and b) set up an event time based timer to fire off at the latest timestamp. 
- Save all the events for a particular key into the ordered list state (with some optimization exceptions) because the events received in this DoFn are per key - they are not guaranteed to be contiguous and can't be processed right away.
- Once the timer fires off there is a guarantee that all the events for a given key have been received up to the firing timestamp. The latest contiguous range stored in the processing state is used to limit the events in the ordered list state that can be safely processed in a loop.

This [high level diagram](https://docs.google.com/presentation/d/10D6zd2CR9cgZIY1lH838ia5s-t3-Yjg43wdYcerWN_U/edit#slide=id.g2f745ae1d7a_0_4857) illustrates the overall approach.

There are dedicated unit tests to cover both per-key and global sequence processing. Please refer to them to understand the details of use cases.

Note that the batch unit tests for global processing don't automatically run under global sequencing. This is due to the apparent incorrectness of the DirectRunner implementation (it is supposed to block processing event processing DoFns until the side input is calculated once). The batch processing tests were successfully run manually using DataflowRunner. Additional work will be needed to either fix the DirectRunner, switch to PrismRunner (when it supports all the primitives used in this transform), or enable test to run a DataflowRunner.

### CODE REVIEW

File: sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/ContiguousSequenceRange.java
Comment (damccorm) [reply sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/ContiguousSequenceRange.java:41]: Actually, I think this is answered elsewhere. In the class level comment, could we add something like: `this is primarily used by the DefaultSequenceCombiner, see that for more detail`?

File: sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/combiner/DefaultSequenceCombiner.java
Comment (damccorm) [inline sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/combiner/DefaultSequenceCombiner.java:43]:
```suggestion
 * <p>Produces the largest {@link ContiguousSequenceRange} of contiguous longs which starts from the initial event
```

Not sure if my framing is better or not, but initially I didn't understand that this was supposed to be the largest possible subsequence which begins at the first event, which made understanding the whole approach challenging.

File: sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/combiner/SequenceRangeAccumulator.java
Comment (damccorm) [inline sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/combiner/SequenceRangeAccumulator.java:61]: When would we have ranges which are below the initial sequence? Do we need this, or should we just throw? I'd expect this to be an error case...

File: sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/combiner/SequenceRangeAccumulator.java
Comment (damccorm) [inline sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/combiner/SequenceRangeAccumulator.java:47]: I'm having a hard time following what this does, could we add a comment here?

I'm understanding it basically as holding a bunch of disjoint sequences: `<lower end of range, <upper end of range, timestamp?>>`, but I don't really follow what the timestamp is. Maybe ranges would be a better name as well?

Comment (damccorm) [reply sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/combiner/SequenceRangeAccumulator.java:47]: I think the timestamp is the highest timestamp of any event we've seen, but I don't really understand why we need it

File: sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/combiner/SequenceRangeAccumulator.java
Comment (damccorm) [inline sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/combiner/SequenceRangeAccumulator.java:67]: Rather than doing all of the logic below, could we construct a `SequenceRangeAccumulator` with a single range (containing this element), and then call `merge`? I think it should be similarly efficient and accomplish the same thing

**Review by damccorm — COMMENTED** [review]
I'm going to try to take this in 2 chunks, starting with the side input pieces.

The code itself looks good so far, thanks for the thorough tests in particular

File: sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/combiner/SequenceRangeAccumulator.java
Comment (slilichenko) [reply sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/combiner/SequenceRangeAccumulator.java:61]: It can be the case of data reprocessing when customers need to re-play a number of messages but the source doesn't support the exact rewinding (e.g. Pub/Sub only allows time based reset to a snapshot). Also, there could be use cases where the initial sequence is not known upfront and is derived from some message data. In that case we are only interested in the range from the starting sequence.

File: sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/combiner/SequenceRangeAccumulator.java
Comment (slilichenko) [reply sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/combiner/SequenceRangeAccumulator.java:47]: This max timestamp is critical for correctness - it will be the time to which the event time domain based timers will be set in the processing ParDos. Beam runners will guarantee that all events up to that time are received by the DoFn, and it means that we have a complete set of messages for a given key for a given range.

Comment (damccorm) [reply sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/combiner/SequenceRangeAccumulator.java:47]: Got it - that makes sense, but is not intuitive (or at least wasn't to me) - could we add a comment here (and maybe consider a more descriptive variable name)

File: sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/GlobalSequenceTracker.java
Comment (damccorm) [inline sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/GlobalSequenceTracker.java:97]: Will this just cause this trigger to fire repeatedly after every element it sees?

Comment (damccorm) [reply sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/GlobalSequenceTracker.java:97]: If so, is there a point in emitting after processing time?

File: sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/GlobalSequencesProcessorDoFn.java
Comment (damccorm) [inline sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/GlobalSequencesProcessorDoFn.java:58]: Hm, noting that I think the usage of this timer will cause issues if the input has allowed lateness. We may need to disallow handling late data here

File: sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/UnprocessedEvent.java
Comment (damccorm) [inline sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/UnprocessedEvent.java:71]: Probably this would be good to add

**Review by damccorm — COMMENTED** [review]
Thanks - mostly LGTM outside of pending comments (mostly cosmetic)

File: sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/ContiguousSequenceRange.java
Comment (slilichenko) [reply sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/ContiguousSequenceRange.java:38]: Thanks - makes sense. I made ContiguousSequenceRange open ended.

File: sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/ContiguousSequenceRange.java
Comment (slilichenko) [reply sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/ContiguousSequenceRange.java:41]: It's not only for the combiner; it's used in the processing DoFn to DLQ the elements below the range.

File: CHANGES.md
Comment (slilichenko) [reply CHANGES.md]: Removed this line from 2.60.0 section. Will add to 2.61.0 in the next PR.

File: sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/combiner/SequenceRangeAccumulator.java
Comment (slilichenko) [reply sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/combiner/SequenceRangeAccumulator.java:73]: Agree on the same outcome. I am afraid that computationally it's going to be more expensive. Requires more lookups, creating a stream and iterator, etc. For high volume processing it can add up.

File: sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/GlobalSequencesProcessorDoFn.java
Comment (slilichenko) [reply sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/GlobalSequencesProcessorDoFn.java:58]: Because this type of timer doesn't honor the lateness?

File: CHANGES.md
Comment (damccorm) [reply CHANGES.md]: If you pull in the latest changes from master we can still add this to changes.md in this pr. It is ok to defer if you'd prefer though

File: sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/combiner/SequenceRangeAccumulator.java
Comment (damccorm) [reply sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/combiner/SequenceRangeAccumulator.java:67]: Ack - maybe we should at least warn? In this case we are intentionally discarding data, which might be unexpected

File: sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/combiner/SequenceRangeAccumulator.java
Comment (damccorm) [reply sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/combiner/SequenceRangeAccumulator.java:73]: Fair enough, SGTM

File: sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/GlobalSequencesProcessorDoFn.java
Comment (damccorm) [reply sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/GlobalSequencesProcessorDoFn.java:58]: Yeah. I guess we could also look at the input allowed lateness and use that to key the timer though.

File: sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/combiner/SequenceRangeAccumulator.java
Comment (slilichenko) [reply sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/combiner/SequenceRangeAccumulator.java:67]: The data is not discarded in this case; it's output to the DLQ output with "duplicate" reason - ProcessorDoFn, line 349 (for some reason can't create a permalink in PR).

File: sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/combiner/SequenceRangeAccumulator.java
Comment (slilichenko) [reply sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/combiner/SequenceRangeAccumulator.java:56]: Added the comments. I kept the variable names the same; let me know please if the comments are not sufficent.

File: sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/GlobalSequenceTracker.java
Comment (slilichenko) [reply sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/GlobalSequenceTracker.java]: Fixed - thank you!

File: sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/GlobalSequenceTracker.java
Comment (slilichenko) [reply sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/GlobalSequenceTracker.java]: I fixed this class - made the number of elements a parameter to the handler (with default of 1000). Unit tests run green.

File: sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/GlobalSequencesProcessorDoFn.java
Comment (slilichenko) [reply sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/GlobalSequencesProcessorDoFn.java:58]: Good point! Added the logic to capture the lateness of the input and add it to the max timestamp.

File: sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/UnprocessedEvent.java
Comment (slilichenko) [reply sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/UnprocessedEvent.java]: Done

File: sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/combiner/SequenceRangeAccumulator.java
Comment (damccorm) [reply sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/combiner/SequenceRangeAccumulator.java:77]: Oh I see, you're right. Thanks

File: sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/combiner/SequenceRangeAccumulator.java
Comment (damccorm) [reply sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/combiner/SequenceRangeAccumulator.java:56]: LGTM, thank you

**Review by damccorm — APPROVED** [review]
This LGTM, thanks! I noticed that the java test suite failed, but it looks unrelated to your change (`FlinkRequiresStableInputTest` failed). I'm rerunning to hopefully get a green signal before merging

**Pull request merged**
